src/Language/PureScript/Sugar/CaseDeclarations.hs

```diff
@@ -343,76 +337,77 @@ inSameGroup (ValueDeclaration vd1) (ValueDeclaration vd2) = valdeclIdent vd1 ==
+getBinders :: (a, [Binder], b) -> [Binder]
+getBinders (_, bs, _) = bs
+
+makeCaseDeclaration :: forall m. (MonadSupply m) => SourceAnn -> Ident -> [(SourceAnn, [Binder], [GuardedExpr])] -> m Declaration
+makeCaseDeclaration sa@(ss, _) ident alternatives = do
+  let namedArgs = map findName . getBinders <$> alternatives
       argNames = foldl1 resolveNames namedArgs
-  args <- if allUnique (catMaybes argNames)
+  args <- if allUnique (catMaybes (map snd argNames))
             then mapM argName argNames
-            else replicateM (length argNames) freshIdent'
-  let vars = map (Var . Qualified Nothing) args
-      binders = [ CaseAlternative bs result | (bs, result) <- alternatives ]
-  let value = foldr (Abs . VarBinder) (Case vars binders) args
+            else zip (map fst argNames) <$> replicateM (length argNames) freshIdent'
+  let vars = map (\(ss', i) -> Var (ss', []) (Qualified Nothing i)) args
+      binders = [ CaseAlternative ss' bs result | ((ss', _), bs, result) <- alternatives ]
+  let value = foldr (\(ss', i) -> Abs (ss', []) (VarBinder ss i)) (Case sa vars binders) args
 
-  return $ ValueDecl (ss, []) ident Public [] [MkUnguarded value]
+  return $ ValueDecl (ss, []) ident Public [] [mkUnguarded value]
   where
   -- We will construct a table of potential names.
   -- VarBinders will become Just _ which is a potential name.
   -- Everything else becomes Nothing, which indicates that we
   -- have to generate a name.
-  findName :: Binder -> Maybe Ident
-  findName (VarBinder name) = Just name
-  findName (PositionedBinder _ _ binder) = findName binder
-  findName _ = Nothing
+  findName :: Binder -> (SourceSpan, Maybe Ident)
+  findName (VarBinder ss' name) = (ss', Just name)
+  findName b = (binderSourceSpan b, Nothing)
 
   -- We still have to make sure the generated names are unique, or else
   -- we will end up constructing an invalid function.
   allUnique :: (Ord a) => [a] -> Bool
   allUnique xs = length xs == length (ordNub xs)
 
-  argName :: Maybe Ident -> m Ident
-  argName (Just name) = return name
-  argName _ = freshIdent'
+  argName :: (SourceSpan, Maybe Ident) -> m (SourceSpan, Ident)
+  argName (ss', Just name) = pure (ss', name)
+  argName (ss', _) = (,) ss' <$> freshIdent'
 
   -- Combine two lists of potential names from two case alternatives
   -- by zipping correspoding columns.
-  resolveNames :: [Maybe Ident] -> [Maybe Ident] -> [Maybe Ident]
+  resolveNames :: [(SourceSpan, Maybe Ident)] -> [(SourceSpan, Maybe Ident)] -> [(SourceSpan, Maybe Ident)]
   resolveNames = zipWith resolveName
 
   -- Resolve a pair of names. VarBinder beats NullBinder, and everything
   -- else results in Nothing.
-  resolveName :: Maybe Ident -> Maybe Ident -> Maybe Ident
-  resolveName (Just a) (Just b)
-    | a == b = Just a
-    | otherwise = Nothing
-  resolveName _ _ = Nothing
+  resolveName :: (SourceSpan, Maybe Ident) -> (SourceSpan, Maybe Ident) -> (SourceSpan, Maybe Ident)
+  resolveName (ss', Just a) (_, Just b)
```

Comment:
I don't know how to pick an appropriate span here, since we're merging two different binders. So for now I just always pick the first span.